Fix breakpad nightly build + add test canary

driazati opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61159
* #61158
* **#60990**

This makes the breakpad build more explicit in its messaging and hints to cmake where to look for the library (it wasn't able to find it without `PATHS` on CI even though that works locally). This also adds a smoke test that will fail if breakpad isn't present on a CI job where it is expected (e.g. binary builds).

Tested on [a nightly wheel](https://app.circleci.com/pipelines/github/pytorch/pytorch/345379/workflows/c68490aa-efba-4788-98fb-874bdb707b26/jobs/14545463/artifacts):

```
curl -LO 'https://14545463-65600975-gh.circle-artifacts.com/0/final_pkgs/torch-1.10.0.dev20210701%2Bcu113-cp39-cp39-linux_x86_64.whl'
pip install torch-1.10.0.dev20210701%2Bcu113-cp39-cp39-linux_x86_64.whl
python
>>> import torch
>>> torch.utils._crash_handler.enable_minidumps()
>>> torch.bincount(input=torch.tensor([9223372036854775807]))
Wrote minidump to /tmp/pytorch_crashes/ad0b88a4-b929-48d4-95818580-d5911f80.dmp
fish: “python” terminated by signal SIGSEGV (Address boundary error)
```

Differential Revision: [D29514316](https://our.internmc.facebook.com/intern/diff/D29514316)

driazati committed:
[test][wip] show the stuff

[ghstack-poisoned]

driazati committed:
Update on "[test][wip] show the stuff"




[ghstack-poisoned]

driazati committed:
Update on "[test][wip] show the stuff"




[ghstack-poisoned]

driazati committed:
Update on "[test][wip] show the stuff"




[ghstack-poisoned]

driazati committed:
Update on "[test][wip] show the stuff"




[ghstack-poisoned]

driazati committed:
Update on "[test][wip] show the stuff"




[ghstack-poisoned]

driazati committed:
Update on "Fix breakpad build + add test canary"


This makes the breakpad build more explicit in its messaging and hints to cmake where to look for the library (it wasn't able to find it without `PATHS` on CI even though that works locally). This also adds a smoke test that will fail if breakpad isn't present on a CI job where it is expected (e.g. binary builds).


[ghstack-poisoned]

driazati committed:
Update on "Fix breakpad build + add test canary"


This makes the breakpad build more explicit in its messaging and hints to cmake where to look for the library (it wasn't able to find it without `PATHS` on CI even though that works locally). This also adds a smoke test that will fail if breakpad isn't present on a CI job where it is expected (e.g. binary builds).


[ghstack-poisoned]

driazati commented:
@driazati has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29514316).

driazati committed:
Update on "Fix breakpad build + add test canary"


This makes the breakpad build more explicit in its messaging and hints to cmake where to look for the library (it wasn't able to find it without `PATHS` on CI even though that works locally). This also adds a smoke test that will fail if breakpad isn't present on a CI job where it is expected (e.g. binary builds).

Differential Revision: [D29514316](https://our.internmc.facebook.com/intern/diff/D29514316)

[ghstack-poisoned]

driazati commented:
@driazati has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29514316).

driazati committed:
Update on "Fix breakpad build + add test canary"


This makes the breakpad build more explicit in its messaging and hints to cmake where to look for the library (it wasn't able to find it without `PATHS` on CI even though that works locally). This also adds a smoke test that will fail if breakpad isn't present on a CI job where it is expected (e.g. binary builds).

Tested on [a nightly wheel](https://app.circleci.com/pipelines/github/pytorch/pytorch/345379/workflows/c68490aa-efba-4788-98fb-874bdb707b26/jobs/14545463/artifacts):

```
curl -LO 'https://14545463-65600975-gh.circle-artifacts.com/0/final_pkgs/torch-1.10.0.dev20210701%2Bcu113-cp39-cp39-linux_x86_64.whl'
pip install torch-1.10.0.dev20210701%2Bcu113-cp39-cp39-linux_x86_64.whl
python
>>> import torch
>>> torch.utils._crash_handler.enable_minidumps()
>>> torch.bincount(input=torch.tensor([9223372036854775807]))
Wrote minidump to /tmp/pytorch_crashes/ad0b88a4-b929-48d4-95818580-d5911f80.dmp
fish: “python” terminated by signal SIGSEGV (Address boundary error)
```

Differential Revision: [D29514316](https://our.internmc.facebook.com/intern/diff/D29514316)

[ghstack-poisoned]

malfet commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2375,3 +2376,22 @@ def disable_gc():
         yield
+
+def has_breakpad():
+    # In CI we expect these builds to have breakpad included, if they don't
+    # then something is probably wrong with the build and we need the test
+    # to fail in order to be alerted
+    job_name = os.getenv("CIRCLE_JOB", os.getenv("BUILD_ENVIRONMENT", ""))
+    prefixes = [
+        "pytorch_linux_bionic_py3_8_gcc9_coverage",
+        "binary_linux_manywheel",
+    ]
+    if any([prefix in job_name for prefix in prefixes]):
+        return True
```
Can you please delete those (we should not hardcode any assumptions about CI in torch/testing folder

malfet commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2375,3 +2376,22 @@ def disable_gc():
     else:
         yield
+
+def has_breakpad():
```
```suggestion
def has_breakpad() -> bool:
```

malfet commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2375,3 +2376,22 @@ def disable_gc():
+        return True
+
+    # If not on a special build, check that the library was actually linked in
+    try:
+        torch._C._get_minidump_directory()  # type: ignore[attr-defined]
+        return True
+    except RuntimeError as e:
+        return "Minidump handler is uninintialized, make sure to call" in str(e)
```
```suggestion
    try:
        torch._C._get_minidump_directory()  # type: ignore[attr-defined]
        return True
    except RuntimeError as e:
        return False
```

driazati committed:
Update on "Fix breakpad build + add test canary"


This makes the breakpad build more explicit in its messaging and hints to cmake where to look for the library (it wasn't able to find it without `PATHS` on CI even though that works locally). This also adds a smoke test that will fail if breakpad isn't present on a CI job where it is expected (e.g. binary builds).

Tested on [a nightly wheel](https://app.circleci.com/pipelines/github/pytorch/pytorch/345379/workflows/c68490aa-efba-4788-98fb-874bdb707b26/jobs/14545463/artifacts):

```
curl -LO 'https://14545463-65600975-gh.circle-artifacts.com/0/final_pkgs/torch-1.10.0.dev20210701%2Bcu113-cp39-cp39-linux_x86_64.whl'
pip install torch-1.10.0.dev20210701%2Bcu113-cp39-cp39-linux_x86_64.whl
python
>>> import torch
>>> torch.utils._crash_handler.enable_minidumps()
>>> torch.bincount(input=torch.tensor([9223372036854775807]))
Wrote minidump to /tmp/pytorch_crashes/ad0b88a4-b929-48d4-95818580-d5911f80.dmp
fish: “python” terminated by signal SIGSEGV (Address boundary error)
```

Differential Revision: [D29514316](https://our.internmc.facebook.com/intern/diff/D29514316)

[ghstack-poisoned]

driazati commented:
@driazati has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29514316).
